src/publisher/aws_events.py

```diff
@@ -1,52 +1,108 @@
+from ordered_set import OrderedSet
 import json
 from django.conf import settings
 import boto3
 from publisher import relation_logic as relationships
-from publisher.utils import lmap
+from functools import wraps
+
+from multiprocessing import SimpleQueue
 
 import logging
 
 LOG = logging.getLogger(__name__)
 
-def sns_topic_arn(**overrides):
+def sns_topic_arn():
     "returns an arn path to an AWS event bus. this is used to connect and send/receive events"
     vals = {}
     vals.update(settings.EVENT_BUS)
-    vals.update(overrides)
     # ll: arn:aws:sns:us-east-1:112634557572:bus-articles--ci
     arn = "arn:aws:sns:{region}:{subscriber}:{name}--{env}".format(**vals)
     LOG.info("using topic arn: %s", arn)
     return arn
 
-def event_bus_conn(**overrides):
+def event_bus_conn():
     sns = boto3.resource('sns')
-    return sns.Topic(sns_topic_arn(**overrides))
+    return sns.Topic(sns_topic_arn())
+
+#
+#
+#
+
+SAFEWORD = START = STOP = 'cacao'
+
+def defer(fn):
+    """calls function normally until safeword is received then buffers all requests until the safeword is called again.
+    when the safeword is received a second time, the wrapped function is called with the UNIQUE set of arguments - i.e. it won't be called with the same arguments twice.
+
+    using `defer` limits function arguments to hashable types only.
+
+    """
+
+    call_queue = SimpleQueue()
+    deferring = False
+
+    @wraps(fn)
+    def wrapper(*args):
+        arg = args[0]
+        nonlocal deferring # stateful!
+
+        if arg == SAFEWORD:
+            deferring = not deferring
+            if deferring:
+                # nothing else to do this turn
+                return
+
+            # we're not deferring and we have stored calls to process
+            if not call_queue.empty():
+                # input order cannot be guaranteed as we're using multiprocessing
+                # single-process input order can be guaranteed
+                calls = OrderedSet()
+                while not call_queue.empty():
+                    calls.add(call_queue.get())
+                return [fn(*fnargs) for fnargs in calls]
+
+            else:
+                # we're not deferring and we have no calls to process
+                # TODO: empty list or None ?
+                return
+
+        # store the args if we're deferring and return
+        if deferring:
+            call_queue.put(args)
+            return
+
+        # we're not deferring, call wrapped fn as normal
+        return fn(*args)
+    return wrapper
+
 
 #
 #
 #
 
-def notify(art, **overrides):
+@defer
+def notify(msid):
```

Comment:
the manuscript ID is an integer. some systems pad it with zeroes, lax doesn't, except when sending it out into the world. if it were padded, I'd have `padded_msid` or similar. I think storing the padded msid is a bad idea and tried to get it changed at the beginning of the 2.0 overhaul but to no avail.